[AutoDiff] Registers VJPs for `FloatingPoint.[maximum|minimum]`

vguerra opened the pull request:
Resolves TF-1134.

vguerra committed:
[AutoDiff] Registers VJPs for `FloatingPoint.[maximum|minimum]`

Resolves TF-1134.

vguerra commented on stdlib/public/Differentiation/FloatingPointDifferentiation.swift.gyb:
```diff
@@ -282,4 +282,28 @@ where
+  static func _vjpMinimum(_ x: Self, _ y: Self) -> (
+    value: Self, pullback: (TangentVector) -> (TangentVector, TangentVector)
+  ) {
+    func pullback(_ v: TangentVector) -> (TangentVector, TangentVector) {
```
This pullback is defined according to the logic of [`FloatingPoint.minimum`](https://github.com/apple/swift/blob/main/stdlib/public/core/FloatingPoint.swift#L1700-L1704) so that the corner cases for returning either `x` or `y` are respected.

rxwei replied:
This implementation is running the identical control flow in the VJP and the pullback. It should be simplified so that the pullback is just straight line code and the control flow is executed only once. You can do:

```swift
@inlinable
@derivative(of: minimum)
static func _vjpMinimum(_ x: Self, _ y: Self) -> (
  value: Self, pullback: (TangentVector) -> (TangentVector, TangentVector)
) {
  if x <= y || y.isNaN { return (x, { v in (v, .zero) }) }
  return (y, { v in (.zero, v) })
}
```

The same suggestion applies to `maximum`.

vguerra replied:
Ah I see.. yes, thanks for the remark @rxwei .. i am fixing that.

vguerra replied:
fixed in bc40409.

rxwei replied:
Thanks.

vguerra commented on stdlib/public/Differentiation/FloatingPointDifferentiation.swift.gyb:
```diff
@@ -282,4 +282,28 @@ where
+  static func _vjpMaximum(_ x: Self, _ y: Self) -> (
+    value: Self, pullback: (TangentVector) -> (TangentVector, TangentVector)
+  ) {
+    func pullback(_ v: TangentVector) -> (TangentVector, TangentVector) {
```
same comment as above but for [`FloatingPoint.maximum`](https://github.com/apple/swift/blob/main/stdlib/public/core/FloatingPoint.swift#L1706-L1710).

dan-zheng reviewed (approved):
IANAM (I am not a mathematician) but LGTM! Let me ping one.

dan-zheng commented:
@swift-ci Please test

vguerra committed:
Simplifying control flow within pullbacks.

To avoid executing control flow code twice ( once in VJP and
once in the pullback ).

rxwei commented:
@swift-ci please test
